src/storage/secondary/block/rle_block_iterator.rs

```diff
@@ -0,0 +1,451 @@
+            cur_array: None,
+        }
+    }
+
+    fn get_cur_rle_count(&self) -> u16 {
+        let mut rle_buffer = &self.rle_block[self.cur_row * std::mem::size_of::<u16>()..];
+        rle_buffer.get_u16_le()
+    }
+}
+
+impl<A, B> BlockIterator<A> for RleBlockIterator<A, B>
+where
+    A: Array,
+    B: BlockIterator<A>,
+{
+    fn next_batch(&mut self, expected_size: Option<usize>, builder: &mut A::Builder) -> usize {
+        if self.cur_row >= self.rle_row_count {
+            return 0;
+        }
+
+        // TODO(chi): error handling on corrupted block
+
+        let mut cnt = 0;
+        // If self.cur_array is none, then we need to get the first array from block_iter
+        // Every time we get only one item from block_iter
+        if self.cur_array.is_none() {
+            let mut array_builder = A::Builder::new();
+            if self.block_iter.next_batch(Some(1), &mut array_builder) == 0 {
+                return cnt;
+            }
+            self.cur_array = Some(array_builder.finish());
+        }
+        let mut rle_count = self.get_cur_rle_count();
+
+        loop {
+            if let Some(expected_size) = expected_size {
+                assert!(expected_size > 0);
+                if cnt >= expected_size {
+                    break;
+                }
+            }
+
+            // Check if we need to get the next array from block_iter
+            if self.cur_rle_scanned < rle_count as usize {
+                builder.append(self.cur_array.as_ref().unwrap());
+                self.cur_rle_scanned += 1;
+                cnt += 1;
+            } else {
+                // Every time cur_row is updated, we need to get the next array from block_iter
+                // And reset cur_rle_scanned
+                self.cur_row += 1;
+                self.cur_rle_scanned = 0;
+                if self.cur_row >= self.rle_row_count {
+                    break;
+                }
+                let mut array_builder = A::Builder::new();
+                if self.block_iter.next_batch(Some(1), &mut array_builder) == 0 {
+                    break;
+                }
+                self.cur_array = Some(array_builder.finish());
+                rle_count = self.get_cur_rle_count();
+            }
+        }
+
+        cnt
+    }
+
+    fn skip(&mut self, cnt: usize) {
+        let mut cnt = cnt;
+        while cnt > 0 {
+            let rle_count = self.get_cur_rle_count();
+            let cur_left = rle_count as usize - self.cur_rle_scanned;
+            if cur_left > cnt {
+                self.cur_rle_scanned += cnt;
+                cnt = 0;
+            } else {
+                cnt -= cur_left;
+                self.cur_rle_scanned = 0;
+                self.cur_row += 1;
+                self.block_iter.skip(1);
+                if self.cur_row >= self.rle_row_count {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn remaining_items(&self) -> usize {
```

Comment:
This remaining_items implementation is too heavy. Is it possible to calculate the total row in advance (e.g., in `new`), and simply return `total_row - current_row_id`?